rfc/0001-policy-server-custom-resource.md

```diff
@@ -0,0 +1,426 @@
+>   * what other parts of the product will be affected?
+>   * will the solution be hard to maintain in the future?
+
+## Maintaining the PolicyServer Custom Resource
+
+Each configuration knob of the Policy Server will have to be exposed
+on the Custom Resource. This is however already happening because, while
+the `policy-server` ConfigMap is kinda "free-form", we still have to
+write code that fetches the right values from the ConfigMap and put them
+into the Deployment definition of Policy Server.
+
+By using a dedicated Custom Resource to define the configuration knobs of
+Policy Server we will provide a better UX to our end user (as opposed to
+writing key-values into a ConfigMap). However, exposing configuration knobs
+will require more thinking (to find the good ergonomic) and more code to
+handle that.
+
+Let's take a look at this snippet. This exposes the contents of the `sources.yaml`
+file used by Policy Server:
+
+```yaml
+  sources:
+    insecure:
+    - insecure1.registry.foo.bar
+    - insecure2.registry.foo.bar
+    secure:
+    - name: self-signed1.registry.com
+      certificate: <base64 blob>
+    - name: self-signed2.registry.com
+      certificateFrom:
+        configMapKeyRef:
+          name: name-of-the-configmap
+          key: ca_key_name
+```
+
+The way to express self-signed registries (the `secure` section) takes
+some inspiration from how Secret/ConfigMap values are mounted into
+Pods. This is definitely more "Kubernetes native", but of course we will
+have to define custom Structs to hold this data, and write more code
+to handle them.
+
+
+## Welcome back, cert-manager?
+
+As discussed before, we must have a mutation webhook to
+process `PolicyServer` and `ClusterAdmissionPolicy`
+resources.
+
+This is extremely important, because we must ensure we have
+all the finalizers set and the user is not allowed harm himself
+(eg: rename a PolicyServer, delete the `default` PolicyServer,...).
+
+The controller itself can act as a mutation webhook, that's a common
+practice with Kubebuilder.
+
+However, the webhook endpoint must be secured with a TLS certificate, which
+has to be somehow provided.
+
+This is usually done by leveraging [cert-manager](https://cert-manager.io/),
+which must be previously installed on the cluster.
+
+Should we just bite the bullet and introduce this external dependency? After all,
+cert-manager is already mandatory to install many other Kubernetes add-ons.
+
+Maybe we can extend our helm chart to have these configuration knobs:
+
+  * Install cert-manager
+  * Do not install cert-manager, because it already exists. I think this
+    should be the recommended option for production deployments
+  * Generate the secrets via helm. This [seems doable](https://medium.com/nuvo-group-tech/move-your-certs-to-helm-4f5f61338aca).
+    We would not be able to rotate them, this is definitely something for quick evaluation
+    of Kubewarden. That's not meant to be used in production
```

Comment:
Another option exists: kubewarden-controller generates this secrets upon start and registers the webhook. It just needs a powerful enough kubeconfig with rights to do some key operations that lead to it registering the webhooks to itself on the cluster where it's running.

The previous option can either be used for non-production and production environments, depending on the trust level of the execution environment etc of kubewarden-controller itself.

I agree that maybe? in the majority of production environments users will prefer to use cert-manager.

Reply:
I've added a note about that.